aws-s3-transfer-manager/src/middleware/hedge.rs

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::time::Duration;
+use tower::{
+    hedge::{Hedge, Policy},
+    layer::layer_fn,
+    BoxError, Layer, Service,
+};
+
+/// S3 recommends retrying the slowest 5% of the requests.
+const LATENCY_PERCENTILE: f32 = 95.0;
+/// This value was chosen randomly and empirically tested.
+const MIN_DATA_POINTS: u64 = 20;
+/// The Hedge layer maintains two rotating histograms, i.e., ReadHistogram and WriteHistogram. They
+/// are switched every period. This value was chosen randomly but with consideration given to the fact that most 8-16 MB part
+/// requests take on average 0.2 seconds, and we should retry it if it takes more than a second.
+const PERIOD: Duration = Duration::new(2, 0);
+
+/*
+* During uploads, S3 recommends retrying the slowest 5% of requests for latency-sensitive applications,
+* as some requests can experience high time to first byte. If a slow part is hit near the end of the request,
+* the application may spend the last few seconds waiting for those final parts to complete, which can reduce overall
+* throughput. This layer is used to retry the slowest 5% of requests to improve performance.
+* Based on our experiments, this makes a significant difference for multipart upload use-cases and
+* does not have a noticeable impact for the Download.
+*/
+pub(crate) struct Builder<P> {
+    policy: P,
+    latency_percentile: f32,
+    min_data_points: u64,
+    period: Duration,
+}
+
+#[derive(Debug, Clone, Default)]
+pub(crate) struct DefaultPolicy;
+
+impl<T: Clone> Policy<T> for DefaultPolicy {
+    fn clone_request(&self, req: &T) -> Option<T> {
+        Some(req.clone())
+    }
```

Comment:
Since this is not a public API we can change it later, but here's my two cents.

If an API takes a reference and internally clones it to do something, it's more honest to the callers of the API if the API takes ownership of the passed-in argument (sort of documented [here](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control)).

Something like
```
    fn clone_request(&self, req: T) -> Option<T> {
        Some(req)
    }
```
At this point, we may or may not need the trait bound `Clone` on `T`.

Curious what you think, though.

Reply:
I don't understand it. The link states that "If a function does not require ownership of an argument, it should take a shared or exclusive borrow of the argument rather than taking ownership and dropping the argument." Since we don't need ownership of `T` here, should we take a reference? 

Also, this trait is not defined by us; it is defined by Tower at https://github.com/tower-rs/tower/blob/master/tower/src/hedge/mod.rs#L58.
```
fn clone_request(&self, req: &Request) -> Option<Request>;
```
So we are getting a reference to Request instead of the ownership.

Reply:
I think what this API is doing is that since the first request is taking too long, we need to send another one. So we clone the request and send the second request. Both requests will stay alive until one of them completes, so we can't drop the original request yet.

Reply:
> this trait is not defined by us

Oh my bad, I don't know where I was looking at, and this is specifically a method for `clone` (for some reason I didn't pay attention to the method name).

The link applies to many cases, but `clone_request` is an exception because of what it does.